frmts/ogcapi/gdalogcapidataset.cpp

```diff
@@ -1670,7 +1670,16 @@ bool OGCAPIDataset::InitWithTilesAPI(GDALOpenInfo *poOpenInfo,
     if (!DownloadJSon(osTilingSchemeURL.c_str(), oDoc, nullptr,
                       MEDIA_TYPE_JSON))
         return false;
-    auto tms = gdal::TileMatrixSet::parse(oDoc.SaveAsString().c_str());
+
+    const auto uri = oDoc.GetRoot().GetString("uri");
+
+    if (uri.empty())
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "Cannot parse TMS uri");
```

Comment:
Not sure what the logic is, but it is totally OK for a 2D TMS to not have a URI, so this should probably not fail here?
Only registered 2DTMS have a `uri`.

Also, I would also expect the URI-matching-based identification (for registered 2DTMSs) to have happened before using the `tileMatrixSetURI` property of the tileset instead. If the `tileMatrixSetURI` was a match for a registered 2DTMS, the client would not need to download the 2DTMS definition at all (the `DownloadJSon()` call above I think).

Reply:
@jerstlouis you are right. There are other ways of parsing the TMS information. I will remove this check

Reply:
@doublebyte1 Yes essentially it could use the URI already checked at:

https://github.com/OSGeo/gdal/blob/master/frmts/ogcapi/gdalogcapidataset.cpp#L1489

```cpp
const auto oTileMatrixSetURI = oTileset.GetString("tileMatrixSetURI");
```

and if this URI is predefined (e.g., _WebMercatorQuad_ or _WorldCRS84Quad_), then it does not need to call `DownloadJSon(osTilingSchemeURL.c_str(), ...`.

This `InitWithTilesAPI()` method is very long though, so it would benefit from being split into different methods each performing one specific high level task.